mirdata/giantsteps_key.py

```diff
@@ -0,0 +1,322 @@
+key changes, and additional modal changes. This dataset is a revision of the original  GiantSteps Key Dataset, available
+in Github (<https://github.com/GiantSteps/giantsteps-key-dataset>) and initially described in:
+
+Knees, P., Faraldo, Á., Herrera, P., Vogl, R., Böck, S., Hörschläger, F., Le Goff, M. (2015).
+Two Datasets for Tempo Estimation and Key Detection in Electronic Dance Music Annotated from User Corrections.
+In Proceedings of the 16th International Society for Music Information Retrieval Conference, 364–370. Málaga, Spain.
+
+The original audio samples belong to online audio snippets from Beatport, an online music store for DJ's and Electronic
+Dance Music Producers (<http:\\www.beatport.com>). If this dataset were used in further research, we would appreciate
+the citation of the current DOI (10.5281/zenodo.1101082) and the following doctoral dissertation, where a detailed
+description of the properties of this dataset can be found:
+
+Ángel Faraldo (2017). Tonality Estimation in Electronic Dance Music: A Computational and Musically Informed Examination.
+PhD Thesis. Universitat Pompeu Fabra, Barcelona.
+
+This dataset is mainly intended to assess the performance of computational key estimation algorithms in electronic dance
+music subgenres.
+
+All the data of this dataset is licensed with Creative Commons Attribution Share Alike 4.0 International.
+
+"""
+
+import csv
+import json
+import zipfile
+
+import librosa
+import numpy as np
+import os
+
+from mirdata import download_utils
+from mirdata import jams_utils
+from mirdata import track
+from mirdata import utils
+
+DATASET_DIR = 'GiantSteps_key'
+REMOTES = {
+    'audio': download_utils.RemoteFileMetadata(
+        filename='audio.zip',
+        url='https://zenodo.org/record/1095691/files/audio.zip?download=1',
+        checksum='8ec9ade888d5a88ce435d7fda031929b',
+        destination_dir='.',
+    ),
+    'keys': download_utils.RemoteFileMetadata(
+        filename='keys.zip',
+        url='https://zenodo.org/record/1095691/files/keys.zip?download=1',
+        checksum='775b7d17e009f5818544cf505b6a96fd',
+        destination_dir='.',
+    ),
+    'metadata': download_utils.RemoteFileMetadata(
+        filename='original_metadata.zip',
+        url='https://zenodo.org/record/1095691/files/original_metadata.zip?download=1',
+        checksum='54181e0f34c35d9720439750d0b08091',
+        destination_dir='.',
+    ),
+}
+
+DATA = utils.LargeData('giantsteps_key_index.json')
+
+
+class Track(track.Track):
+    """giantsteps_key track class
+
+    Args:
+        track_id (str): track id of the track
+        data_home (str): Local path where the dataset is stored.
+            If `None`, looks for the data in the default directory, `~/mir_datasets`
+
+    Attributes:
+        audio_path (str): track audio path
+        keys_path (str): key annotation path
+        metadata_path (str): sections annotation path
```

Comment:
What's in the metadata file? In most of the other data loaders, we're expanding out any metadata like this into individual attributes so that it's easier to understand what exactly is in the dataset. For an example, see how it's done in `dali.py` with `self._track_metadata`

Reply:
ok perfect

Reply:
This is an example of metadata:
```JSON
{
 "exclusive": false, 
 "images": {
  "large": {
   "url": "https://geo-media.beatport.com/image/8912740.jpg", 
   "width": 500, 
   "id": 8912740, 
   "height": 500
  }, 
  "small": {
   "url": "https://geo-media.beatport.com/image/2489.jpg", 
   "width": 30, 
   "id": 2489, 
   "height": 30
  }, 
  "medium": {
   "url": "https://geo-media.beatport.com/image/276.jpg", 
   "width": 60, 
   "id": 276, 
   "height": 60
  }, 
  "dynamic": {
   "url": "https://geo-media.beatport.com/image_size{hq}/{w}x{h}/8912740.jpg", 
   "id": 8912740
  }
 }, 
 "artists": [
  {
   "slug": "jason-sparks", 
   "id": 681, 
   "name": "Jason Sparks"
  }
 ], 
 "duration": {
  "minutes": "3:10", 
  "milliseconds": 190840
 }, 
 "id": 10089, 
 "genres": [
  {
   "slug": "breaks", 
   "id": 9, 
   "name": "Breaks"
  }
 ], 
 "title": "Close My Eyes feat. J. Little (Original Mix)", 
 "label": {
  "slug": "botchit-and-scarper", 
  "id": 85, 
  "name": "Botchit & Scarper"
 }, 
 "mix": "Original Mix", 
 "preview": {
  "mp4": {
   "url": "https://geo-samples.beatport.com/lofi/10089.LOFI.mp4", 
   "offset": {
    "start": 70840, 
    "end": 190840
   }
  }, 
  "mp3": {
   "url": "https://geo-samples.beatport.com/lofi/10089.LOFI.mp3", 
   "offset": {
    "start": 70840, 
    "end": 190840
   }
  }
 }, 
 "type": "track", 
 "remixers": [], 
 "purchase_type": null, 
 "audio_format": "mp3", 
 "sale_type": "purchase", 
 "purchase": 1, 
 "price": {
  "symbol": "\u20ac", 
  "code": "EUR", 
  "display": "\u20ac1.30", 
  "value": 1.3
 }, 
 "component": "Track Detail", 
 "sponsored": false, 
 "key": "D maj", 
 "date": {
  "released": "2004-02-23", 
  "published": "2004-02-23"
 }, 
 "active": true, 
 "slug": "close-my-eyes-feat-j-little-original-mix", 
 "preorder": false, 
 "name": "Close My Eyes feat. J. Little", 
 "component_type": null, 
 "bpm": 150, 
 "formats": {
  "wav": {
   "symbol": "\u20ac", 
   "code": "EUR", 
   "display": "\u20ac0.75", 
   "value": 0.75
  }, 
  "aiff": {
   "symbol": "\u20ac", 
   "code": "EUR", 
   "display": "\u20ac0.75", 
   "value": 0.75
  }
 }, 
 "release": {
  "slug": "heroes-and-villians-ep", 
  "id": 1449, 
  "name": "Heroes & Villians EP"
 }, 
 "waveform": {
  "large": {
   "url": "https://geo-media.beatport.com/image/8912747.png", 
   "width": 1500, 
   "id": 8912747, 
   "height": 250
  }, 
  "dynamic": {
   "url": "https://geo-media.beatport.com/image_size{hq}/{w}x{h}/8912747.png", 
   "id": 8912747
  }
 }, 
 "guest_pick": false, 
 "sub_genres": []
}
```

Maybe this part is not needed.
```JSON
"images": {
  "large": {
   "url": "https://geo-media.beatport.com/image/8912740.jpg", 
   "width": 500, 
   "id": 8912740, 
   "height": 500
  }, 
  "small": {
   "url": "https://geo-media.beatport.com/image/2489.jpg", 
   "width": 30, 
   "id": 2489, 
   "height": 30
  }, 
  "medium": {
   "url": "https://geo-media.beatport.com/image/276.jpg", 
   "width": 60, 
   "id": 276, 
   "height": 60
  }, 
  "dynamic": {
   "url": "https://geo-media.beatport.com/image_size{hq}/{w}x{h}/8912740.jpg", 
   "id": 8912740
  }
 }, 

```
I don't know if the rest of the information is important enough to make the loader more complex. Perhaps it would be better if I explain in the docstring each parameter of the metadata in detail. What do you think @rabitt ?

Reply:
hey, @rabitt My proposal is this one:
```
{
 // "exclusive": false, 
 // "images": {
 //  "large": {
 //   "url": "https://geo-media.beatport.com/image/8912740.jpg", 
 //   "width": 500, 
 //   "id": 8912740, 
 //   "height": 500
 //  }, 
 //  "small": {
 //   "url": "https://geo-media.beatport.com/image/2489.jpg", 
 //   "width": 30, 
 //   "id": 2489, 
 //   "height": 30
 //  }, 
 //  "medium": {
 //   "url": "https://geo-media.beatport.com/image/276.jpg", 
 //   "width": 60, 
 //   "id": 276, 
 //   "height": 60
 //  }, 
 //  "dynamic": {
 //   "url": "https://geo-media.beatport.com/image_size{hq}/{w}x{h}/8912740.jpg", 
 //   "id": 8912740
 //  }
 }, 
 "artists": [
  {
   // "slug": "jason-sparks", 
   "id": 681, 
   "name": "Jason Sparks"
  }
 ], 
 "duration": {
  "minutes": "3:10", 
  "milliseconds": 190840
 }, 
 // "id": 10089, 
 "genres": [
  {
   "slug": "breaks", 
   "id": 9, 
   "name": "Breaks"
  }
 ], 
 "title": "Close My Eyes feat. J. Little (Original Mix)", 
 // "label": {
 //  "slug": "botchit-and-scarper", 
 //  "id": 85, 
 //  "name": "Botchit & Scarper"
 // }, 
 // "mix": "Original Mix", 
 // "preview": {
 //  "mp4": {
 //   "url": "https://geo-samples.beatport.com/lofi/10089.LOFI.mp4", 
 //   "offset": {
 //    "start": 70840, 
 //    "end": 190840
 //   }
 //  }, 
 //  "mp3": {
 //   "url": "https://geo-samples.beatport.com/lofi/10089.LOFI.mp3", 
 //   "offset": {
 //    "start": 70840, 
 //    "end": 190840
 //   }
 //  }
 // }, 
 // "type": "track", 
 // "remixers": [], 
 // "purchase_type": null, 
 // "audio_format": "mp3", 
 // "sale_type": "purchase", 
 // "purchase": 1, 
 // "price": {
 //  "symbol": "\u20ac", 
 //  "code": "EUR", 
 //  "display": "\u20ac1.30", 
 //  "value": 1.3
 // }, 
 // "component": "Track Detail", 
 // "sponsored": false, 
 // "key": "D maj", 
 // "date": {
 //  "released": "2004-02-23", 
 //  "published": "2004-02-23"
 // }, 
 // "active": true, 
 // "slug": "close-my-eyes-feat-j-little-original-mix", 
 // "preorder": false, 
 // "name": "Close My Eyes feat. J. Little", 
 // "component_type": null, 
 "bpm": 150, 
 // "formats": {
 //  "wav": {
 //   "symbol": "\u20ac", 
 //   "code": "EUR", 
 //   "display": "\u20ac0.75", 
 //   "value": 0.75
 //  }, 
 //  "aiff": {
 //   "symbol": "\u20ac", 
 //   "code": "EUR", 
 //   "display": "\u20ac0.75", 
 //   "value": 0.75
 //  }
 // }, 
 // "release": {
 //  "slug": "heroes-and-villians-ep", 
 //  "id": 1449, 
 //  "name": "Heroes & Villians EP"
 // }, 
 // "waveform": {
 //  "large": {
 //   "url": "https://geo-media.beatport.com/image/8912747.png", 
 //   "width": 1500, 
 //   "id": 8912747, 
 //   "height": 250
 //  }, 
 //  "dynamic": {
 //   "url": "https://geo-media.beatport.com/image_size{hq}/{w}x{h}/8912747.png", 
 //   "id": 8912747
 //  }
 // }, 
 // "guest_pick": false, 
 "sub_genres": []
}
```

4 new methods


genres and subgenres loader
tempo (annotated by beatports community)
duration
artist

Reply:
ok done in ddd0520

Reply:
Great, I like it!